cpp/src/utilities/device_atomics.cuh

```diff
@@ -0,0 +1,528 @@
+
+            return type_reinterpret<T, T_int>(ret);
+        }
+    };
+
+    // 8 bytes atomic operation
+    template<typename T>
+    struct typesAtomicCASImpl<T, 8> {
+        __forceinline__  __device__
+        T operator()(T* addr, T const & compare, T const & update_value)
+        {
+            using T_int = unsigned long long int;
+
+            T_int ret = atomicCAS(
+                reinterpret_cast<T_int*>(addr),
+                type_reinterpret<T_int, T>(compare),
+                type_reinterpret<T_int, T>(update_value));
+
+            return type_reinterpret<T, T_int>(ret);
+        }
+    };
+
+    // call atomic function with type cast between same underlying type
+    template <typename T, typename Functor>
+    __forceinline__  __device__
+    T typesAtomicOperation32(T* addr, T val, Functor atomicFunc)
+    {
+        using T_int = int;
+        T_int ret = atomicFunc(reinterpret_cast<T_int*>(addr),
+            cudf::detail::type_reinterpret<T_int, T>(val));
+
+        return cudf::detail::type_reinterpret<T, T_int>(ret);
+    }
+
+    // call atomic function with type cast between same underlying type
+    template <typename T, typename Functor>
+    __forceinline__  __device__
+    T typesAtomicOperation64(T* addr, T val, Functor atomicFunc)
+    {
+        using T_int = long long int;
+        T_int ret = atomicFunc(reinterpret_cast<T_int*>(addr),
+            cudf::detail::type_reinterpret<T_int, T>(val));
+
+        return cudf::detail::type_reinterpret<T, T_int>(ret);
+    }
+
+} // namespace detail
+
+
+template <typename T, typename BinaryOp>
+__forceinline__  __device__
+T genericAtomicOperation(T* address, T const & update_value, BinaryOp op)
+{
+    return cudf::detail::genericAtomicOperationImpl<T, BinaryOp, sizeof(T)>()
+        (address, update_value, op);
+}
+
+// ------------------------------------------------------------------------
+// Binary ops for sum, min, max
+struct DeviceSum {
+    template<typename T>
+    __device__
+    T operator() (const T &lhs, const T &rhs) {
+        return lhs + rhs;
+    }
+
+    template<typename T>
+    static constexpr T identity() { return T{0}; }
+};
+
+struct DeviceMin{
+    template<typename T>
+    __device__
+    T operator() (const T &lhs, const T &rhs) {
+        return lhs <= rhs? lhs: rhs;
+    }
+
+    template<typename T>
+    static constexpr T identity() { return std::numeric_limits<T>::max(); }
+};
+
+struct DeviceMax{
+    template<typename T>
+    __device__
+    T operator() (const T &lhs, const T &rhs) {
+        return lhs >= rhs? lhs: rhs;
+    }
+
+    template<typename T>
+    static constexpr T identity() { return std::numeric_limits<T>::lowest(); }
+};
+
+} // namespace cudf
+
+// ------------------------------------------------------------------------
+// ------------------------------------------------------------------------
+// Overloads for `atomicAdd`.
+// Cuda supports `sint32`, `uint32`, `uint64`, `float`, `double`
+
```

Comment:
I think in this section each `atomic*` function that is part of the "public" (even though it's internal to libcudf) interface of this file should have a doxygen comment block.  This way they will end up in the HTML documentation for the library.

Reply:
Add doxygen text for each `atomic*` overloads in the latest commit.